### DISCUSSION TITLE
Jupyterlite in Quepid (Search Relevance Workbeach)

### DISCUSSION THREAD

**Discussion by epugh** [jupyterlite/jupyterlite#778] [Show and tell]
I've just rolled out the first of cut of adding Jupyterlite to Quepid.  Quepid is a Ruby on Rails based tool for collecting human ratings for search queries, and then being a safe environment for a Relevance Engineer to tweak the search engine algorithms.   

Quepid supports classical Information Retrieval metrics like AP, DCG, nDCG.   While we have a few charts that ship as part of Quepid for showing those metrics, I find that often I want to go deeper and really play with the data.   

Jupyterlite gives an easy way for people to interact with the data that is available from Quepid via it's API's.  For example, here we are plotting the number of queries that return specific scores of quality.  The X axis is the quality from 0 to 1, 1 being best:

![image (4)](https://user-images.githubusercontent.com/22395/185749128-69feb6a8-5905-4381-b7c0-b7cae366bd6c.png)

You can try out the (free!) hosted version of Quepid by going to http://app.quepid.com, sign up, and you'll see a Notebooks link in the header UI that pops you to Jupyterlite.   We also provide a Docker image for on-prem users, however I haven't released that version with Jupyterlite rolled in quite yet.

I would love some feedback on how I rolled Jupyterlite in.   The PR for adding it is at https://github.com/o19s/quepid/pull/544.

1. Thoughts on how I integrated it?   I need to support deployment via Docker image and Heroku.   I didn't want to check a million files into my Github repo.  So, my https://github.com/o19s/quepid/blob/master/Dockerfile.dev#L32 get dependencies, and then I run https://github.com/o19s/quepid/blob/master/bin/setup_jupyterlite script to redo my setup.  This does the build process, and then makes a archive file notebooks.gz.   I then check in that file into Github.   As part of the deploy process on Heroku, those files get unzipped into where Rails would like to serve up the files, `./public/notebooks`.   This means any time I want to update Jupyterlite, either due to  a new release, or I tweak my sample Notebooks, I have to rebuild the archive and check it back in, a bit clunky.   I'd love to have two archives, one for the base Jupyterlite, that only gets updated when new releases comes out, and then seperately merge in my files: https://github.com/o19s/quepid/tree/master/jupyterlite/files

2. I tried to get the WebRTC working, and didn't have any luck.  It felt like I wasn't able to actually enable it...  The configuration I updated was https://github.com/o19s/quepid/blob/master/jupyterlite/jupyter-lite.json.

3. I don't want users to change the example notebooks I ship as part of Quepid, I want them to copy and clone them for their own data.  Any thoughts on how to encourage that, or prevent them from changing the base notebooks?   Future releases of Quepid will ship more example notebooks beyond the one we have today.

4. Can I change the logo without a big theme effort?  I'd love to blend the Quepid and Jupyterlite logo's so folks know which tabs belong to Quepid ;-)

5. Can I generate a link to a notebook that passes parameters in that the notebook would use?   

6. What am I missing that I should take advantage of?

**Comment by jtpio** [comment]
Thanks @epugh for sharing this!

> I tried to get the WebRTC working, and didn't have any luck. It felt like I wasn't able to actually enable it... The configuration I updated was https://github.com/o19s/quepid/blob/master/jupyterlite/jupyter-lite.json.

Did you also install the `jupyterlab-webrtc-docprovider` extension? There is some information about this in https://jupyterlite.readthedocs.io/en/latest/howto/configure/rtc.html

> Any thoughts on how to encourage that, or prevent them from changing the base notebooks?

By default users can edit the notebooks but the changes will be local to them only (saved in the browser). You can also disable persistence with the volatile memory storage: https://jupyterlite.readthedocs.io/en/latest/howto/configure/storage.html#volatile-memory-storage

> Can I change the logo without a big theme effort? I'd love to blend the Quepid and Jupyterlite logo's so folks know which tabs belong to Quepid ;-)

Yes, for example with a plugin like this: https://github.com/jtpio/p5-notebook/blob/5742a5f2ffcb01ca7289be1128dbfca5e7a3ae91/packages/lab-extension/src/index.ts#L15-L57
And disabling the default one with `disabledExtensions` in the config: https://github.com/jtpio/p5-notebook/blob/5742a5f2ffcb01ca7289be1128dbfca5e7a3ae91/jupyter-lite.json#L10
There might be other examples around as well.

> Can I generate a link to a notebook that passes parameters in that the notebook would use?

A link to a notebook yes with the `?path=` URL parameter. More specific parameters not yet, although this might be coming in https://github.com/jupyterlite/jupyterlite/pull/525

**Comment by bollwyvl** [comment]
> Thoughts on how I integrated it?

yeah, dunno, if it's working, probably not going to follow the links. i can only recommend for a "production" deployment of this still very-beta-tool, working with Other Peoples' Data: lock down and "own" as much as you can, as described in the [offline docs page](https://jupyterlite.readthedocs.io/en/latest/howto/configure/advanced/offline.html). various folk have issues with the default CDN posture, which is pretty much appropriate for open source docs sites, but not a whole lot else.

> I want them to copy and clone them for their own data

Lite inherits the same behavior as jupyterlab around read-only files. Marking the files as "read-only" will force this behavior: they can be read and run, and edited, but on save, triggers a save-as dialog. You can do this for real by marking them as such before the build with `chmod` in a script (or your file browser if doing this by hand), or after the fact by modifying `api/contents/*/all.json`.

> a link to a notebook that passes parameters

However, and related to some of your other questions: a forthcoming [jupyterlab-starters](https://github.com/deathbeds/jupyterlab-starters) version will support defining starters in `overrides.json`, and [work with lite](https://jupyterstarters.readthedocs.io/en/latest/demo/index.html?starter=todo). That link shows jumping to a specific starter form, but it wouldn't be too much of a lift to add a pre-seeded body. So if the template for your notebook was:

>  know which tabs belong to Quepid

You'll want to replace the `favicon.ico`, as done on [ipydrawio's](https://github.com/deathbeds/ipydrawio/blob/master/demo/jupyter-lite.json#L5) [lite demo](https://ipydrawio.readthedocs.io/en/stable/_static/lab/index.html).

> missing that I should take advantage of?

If you skimmed the docs, you've probably seen most of it. there isn't a lot of marketing behind lite, and it's mostly been built by and for people who use it for things. I know of [this talk](https://www.youtube.com/watch?v=VII2OL8O4hA) which pushed some of the embedding boundaries at the time, but already seems kind of out-of-date given some of the stuff that was shipped recently (unified contents, etc.)

- have a look further look through the documentation
- do some searching on github for [well-known](https://github.com/search?p=3&q=filename%3Ajupyter_lite_config.json&type=Code) [filenames](https://github.com/search?p=3&q=filename%3Ajupyter-lite.json&type=Code) than might be interesting
- look [on pypi](https://pypi.org/search/?q=&o=&c=Framework+%3A%3A+Jupyter+%3A%3A+JupyterLab+%3A%3A+Extensions+%3A%3A+Prebuilt) for prebuilt jupyterlab extensions
  - there's a chance these will work on 
- look [on pypi](https://pypi.org/search/?q=jupyterlite) for jupyterlite related tools
  - some of these will have very hard-pinned jupyterlite versions, so may not work together, but you can always PR updates to make them work

**Reply by epugh** [reply]
This is some really great input, thank you!   One thing I noticed in the docs is that there is a assumption (and probably a fair assumption) that you already know your way around jupyter first ;-)     I will share back my results from these suggestions!

**Comment by epugh** [comment]
Finally released the Docker image of Quepid ;-)  https://twitter.com/quepid_search/status/1586056610102779906
